### PR TITLE
[1.12] Add Advancement Event

### PR DESCRIPTION
**Pull request by Lomeli12**
Replaces #4257, as that one is seemingly abandoned and includes a ton of unnecessary patches.

Adds the AdvancementEvent, which is triggered when a player recieves an advancement. 

The included test mod, when enabled, simply prints out which player got which advancement to the logs if the advancement would be annouced to the server.

This PR is simply `Post` event from #4257 and does not include a `Pre`. A cancelable `Pre` isn't really necessary when you can simply add more criterion, or add a custom criterion trigger if you need something more complex.

### CODE REVIEW

**Comment by BillyGalbreath** [comment]
I'm glad someone replaced that other pr. I was going to, but when I mentioned it was abandoned I got shunned. *shrugs*

File: src/test/java/net/minecraftforge/debug/AdvancementEventTest.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+        if (ENABLED)
+        {
+            MinecraftForge.EVENT_BUS.register(AdvancementEventTest.class);
+        }
+    }
+
+    @SubscribeEvent
+    public static void onAdvancementEvent(AdvancementEvent event)
+    {
+        if (ENABLED && event.getAdvancement().getDisplay() != null && event.getAdvancement().getDisplay().shouldAnnounceToChat())
```
Comment (liach) [inline src/test/java/net/minecraftforge/debug/AdvancementEventTest.java]: Second `ENABLED` check is not necessary if you have the first one.

File: src/main/java/net/minecraftforge/event/entity/player/AdvancementEvent.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+package net.minecraftforge.event.entity.player;
+
+import net.minecraft.advancements.Advancement;
+import net.minecraft.entity.player.EntityPlayer;
+import net.minecraftforge.common.MinecraftForge;
+import net.minecraftforge.fml.common.eventhandler.Cancelable;
+
+/**
+ * This event is fired when a player gets an advancement
```
Comment (liach) [inline src/main/java/net/minecraftforge/event/entity/player/AdvancementEvent.java]: You might consider adding a period after the sentence.

**Review by Lazarix — APPROVED** [review]

**Comment by LemmaEOF** [comment]
I hope this can get merged in soon. It seems like a pretty stable way to do it and has the approval of a good couple folks. I'm working with advancements for progression right now so it'd be really helpful if I could have these events.

**Pull request merged**